auth/api/iam/openid4vp.go

```diff
@@ -158,6 +163,165 @@ func (r Wrapper) handleAuthorizeRequestFromHolder(ctx context.Context, verifier
+	// missing or invalid parameters are all mapped to invalid_request
+	// any operation that fails is mapped to server_error, this includes unreachable or broken backends.
+
+	// check the response URL because later errors will redirect to this URL
+	responseURI, responseOK := params[responseURIParam]
+
+	// get the original authorization request of the client, if something fails we need the redirectURI from this request
+	// get the state parameter
+	state, ok := params[stateParam]
+	if !ok {
+		// post error to responseURI, if it fails, it'll render error page
+		return r.sendAndHandleDirectPostError(ctx, oauthError(oauth.InvalidRequest, "missing state parameter", nil), responseURI)
+	}
+
+	// check client state
+	// if no state, post error
+	var session OAuthSession
+	err := r.oauthClientStateStore().Get(state, &session)
+	if err != nil {
+		if !responseOK {
+			return nil, oauthError(oauth.ServerError, "something went wrong", nil)
+		}
+		return r.sendAndHandleDirectPostError(ctx, oauthError(oauth.InvalidRequest, "state has expired", nil), responseURI)
+	}
+	clientRedirectURL := session.redirectURI()
+	if !responseOK {
+		if clientRedirectURL != nil {
+			return r.sendAndHandleDirectPostError(ctx, oauthError(oauth.InvalidRequest, "missing response_uri parameter", nil), clientRedirectURL.String())
+		}
+		return nil, oauthError(oauth.ServerError, "something went wrong", nil)
+	}
+
+	verifierID, ok := params[clientIDParam]
+	if !ok {
+		return r.sendAndHandleDirectPostError(ctx, oauthError(oauth.InvalidRequest, "missing client_id parameter", clientRedirectURL), responseURI)
+	}
+	// the verifier must be a did:web
+	verifierDID, err := did.ParseDID(verifierID)
+	if err != nil || verifierDID.Method != "web" {
+		return r.sendAndHandleDirectPostError(ctx, oauthError(oauth.InvalidRequest, "invalid client_id parameter (only did:web is supported)", clientRedirectURL), responseURI)
+	}
+	clientIDScheme, ok := params[clientIDSchemeParam]
+	if !ok || clientIDScheme != didScheme {
+		return r.sendAndHandleDirectPostError(ctx, oauthError(oauth.InvalidRequest, "invalid client_id_scheme parameter", clientRedirectURL), responseURI)
+	}
+	nonce, ok := params[nonceParam]
+	if !ok {
+		return r.sendAndHandleDirectPostError(ctx, oauthError(oauth.InvalidRequest, "missing nonce parameter", clientRedirectURL), responseURI)
+	}
+	// get verifier metadata
+	clientMetadataURI, ok := params[clientMetadataURIParam]
+	if !ok {
+		return r.sendAndHandleDirectPostError(ctx, oauthError(oauth.InvalidRequest, "missing client_metadata_uri parameter", clientRedirectURL), responseURI)
+	}
+	// we ignore any client_metadata, but officially an error must be returned when that param is present.
+	metadata, err := r.auth.Holder().ClientMetadata(ctx, clientMetadataURI)
+	if err != nil {
+		return r.sendAndHandleDirectPostError(ctx, oauthError(oauth.ServerError, "failed to get client metadata (verifier)", clientRedirectURL), responseURI)
+	}
```

Comment:
There are multiple options for various parameters. Let's focus on the happy flow using a single set. This will already make node2node communication via openid4vp possible. Later we can add the other options to be compatible with other implementations.